man/RNAcofold.1.in
.\" .ER
.TH "RNACOFOLD" "l" "1.6" "Ivo Hofacker" "ViennaRNA"
.SH "NAME"
RNAcofold \- calculate secondary structures of two RNAs with dimerization
.SH "SYNOPSIS"
\fBRNAcofold\fP [\-p[\fI0\fP]] [\-C] [\-T\ \fItemp\fP] [\-4] [\-d[0|1|2|3]]
[\-noLP] [\-noGU] [\-noCloseGU] [\-e\ \fI1|2\fP] [\-P\ \fIparamfile\fP]
[\-nsp\ \fIpairs\fP] [\-S\ \fIscale\fP] [\-a\] [\-c\] [\-f\ \fIconcentrationfile\fP]

.SH "DESCRIPTION"
.I RNAcofold
works much like RNAfold, but allows to specify two RNA sequences wich are
then allowed to form a dimer structure. RNA sequences are read from stdin
in the usual format, i.e. each line of input corresponds to one sequence,
except for lines starting with ">" which contain the name of the next 
sequence.
To compute the hybrid structure of two molecules, the
two sequences must be concatenated using the '&' character as separator.
RNAcofold can compute minimum free energy (mfe) structures, as well as
partition function (pf) and base pairing probability matrix (using the
-p switch). 
.br
Since dimer formation is concentration dependent, RNAcofold can be used
to compute equilibrium concentrations for all five monomer and
(homo/hetero)-dimer species, given input concentrations for the monomers.
.br
Output consists of the mfe structure in bracket notation as well as PostScript
structure plots and "dot plot" files containing the pair probabilities, see
the RNAfold man page for details. In the dot plots a cross marks the chain
break between the two concatenated sequences. 
.br 
The program will continue to read new sequences until a line consisting
of the single character @ or an end of file condition is encountered.
.SH "OPTIONS"
.TP
.B \-a
Compute the partition function and free energies not only of the
hetero-dimer consisting of the two input sequences (the "AB dimer"), 
but also of the homo-dimers AA and BB as well as A and B monomers.
The output will contain the free energies for each of these species,
as well as 5 dot plots containing the conditional pair probabilities,
called ABname5.ps, AAname5.ps and so on. For later use, these dot plot
files also contain the free energy of the ensemble as a comment. Using
\fB\-a\fP automatically toggles the 
\fB\-p\fP option.  
.TP
.B \-c
In addition to everything listed under the \fB\-a\fP option, read in
initial monomer concentrations and compute the expected equilibrium
concentrations of the 5 possible species (AB, AA, BB, A, B). Start
concentrations are read from stdin (unless the -f option is
used) in [mol/l], equilibrium concentrations are given realtive to the
sum of the two inputs. An arbitrary number of initial concentrations
can be specified (one pair of concentrations per line).
.TP 
.B \-f\ \fIconcfile\fP
Specify a file with initial concentrations for the to sequences. The table
consits of arbitrary many lines with just two numbers (the concentration of
sequence A and B). Will automatically toggle the \fB\-c\fP (and thus
\fB\-a\fP and \fB\-p\fP) options (see above).
.PP
the \-p, \-p0, \-noGU, \-noCloseGU, \-nsp, \-S, \-e, \-P, \-noLP, \-T, \-4,
\-d and \-C options work as in \fBRNAfold\fP, see there for description.

.SH "REFERENCES"
The energy parameters are taken from:
.br 
D.H. Mathews, J. Sabina, M. Zuker and H. Turner
"Expanded Sequence Dependence of Thermodynamic Parameters Provides 
Robust Prediction of RNA Secondary Structure"
JMB, 288, pp 911\-940, 1999
.PP 
If you use this program in your work you might want to cite:
.PP 
I.L. Hofacker, W. Fontana, P.F. Stadler, S. Bonhoeffer, M. Tacker, P. Schuster 
(1994)
Fast Folding and Comparison of RNA Secondary Structures.
Monatshefte f. Chemie 125: 167\-188
.PP
S.H.Bernhart, Ch. Flamm, P.F. Stadler, I.L. Hofacker
Partition Function and Base Pairing Probabilities of RNA Heterodimers
Algorithms Mol. Biol. (2006)
.SH "VERSION"
This man page documents version @VERSION@ Vienna RNA Package.
.SH "AUTHORS"
Ivo L Hofacker, Peter F Stadler, Stephan Bernhart.
.SH "BUGS"
If in doubt our program is right, nature is at fault.
Comments should be sent to rna@tbi.univie.ac.at.


NEWS
See Changelog for details.

Upcoming 
-----------
- Updated parameters for both RNA and DNA using the data compiled by 
  David Mathews <David_Mathews@urmc.rochester.edu>. 

Version 1.6
------------
- The RNAforester program for tree-alignments of RNA structures is now
  distributed with the Vienna RNA package, see the RNAforester
  subdirectory for more information. RNAforester was written by Matthias
  Hoechsmann <mhoechsm@techfak.uni-bielefeld.de> 
- The Kinfold program for stochastic simulation of folding trajectories is
  now included in the package, see the Kinfold subdirectory.
- cofolding of two structures now supports suboptimal folding and 
  partition function folding (still experimental). 
  ATTENTION: Energies of hybrid structures now inlcude the Duplex-initiation
  energy, which was neglected in previous version.
- RNAplfold is a partition function variant of RNALfold. It computes the
  mean probability of a (local) base pair averaged over all sequence
  windows that contain the pair.
- new utilities to color alignments and consensus structures  
- RNAfold -p now computes the centroid structure
- CAUTION: ensemble diversities in version <1.6.5 are off by a factor 2

Version 1.5pre 
-----------
- ViennaRNA now uses autoconfig generated configure scripts for even better
  portability (should compile on any UNIX, Linux, MacOS X, Windows with
  Cygwin).
- The new RNAalifold program predicts consensus structures for a set of
  aligned sequences.
- Complete suboptimal folding is now integrated in the library.
- Beginning support for co-folding of two strands: energy_of_struct() and
  RNAeval can now compute energies of duplex structures.
- RNAcofold predicts hybrid structures of two RNA strands
- RNAduplex predicts hybrid structures, while allowing only inter-molecular
  base pairs (useful for finding potential binding sites)
- RNALfold predicts locally stable structures in long sequences.
- Major changes to Perl module. See the pod documentation (perldoc RNA).
- RNAsubopt can do stochastic backtracking to produce samples of suboptimal
  structures with Boltzmann statistics.
- New utilities to rotate secondary structure plots and annotate them with
  reliability data.
- Various small bug fixes

Version 1.4
-----------
- New Turner parameters as described in Mathews et.al. JMB v288, 1999.
  Small changes to format of parameter files (old param files won't work!)
- mfe and suboptimal folding will produce only structures without isolated
  pairs if noLonelyPairs=1 (-noLP option), for partition function folding
  pairs that can only occur as isolated pairs are not formed.
- setting dangles=3 (-d3 option) will allow co-axial stacking of adjacent
  helices in mfe folding and energy_of_struct().

Version 1.3.1
-------------
- RNAheat would produce spikes in the specific heat because dangling end
  energies did not go smoothly to 0. 
- PS dot plots now have an option to use a log scale (edit _dp.ps file
  and set logscale to true).

Version 1.3
-----------
- Secondary structure plots now use E. Bruccoleri's naview routines for
  layout by default. New utility RNAplot produces secondary structure plots 
  from structures in bracket notation with several options.
- New -d2 option in RNAfold and RNAeval sets dangles=2, which makes 
  energy_of_struct() and fold() treat dangling ends as in pf_fold().
  -noLP option in RNAfold etc sets noLonelyPairs=1, which avoids most
  structures containing lonely base pairs (helices of length 1).
- new utility functions pack_structure() unpack_structure() make_pair_table()
  and bp_distance().  RNAdistance adds bp_distance() via -DP switch.
- First release of RNAsubopt for complete suboptimal folding.
- fixed bug in asymmetry penalty for interior loops. 
- Default compilation now uses doubles for partition function folding.

Version 1.2.1
-------------
- Fixed bug in version 1.2 of the RNAheat program causing overflow errors
  for most input sequences. 
- The PS_dot_plot() and PS_rna_plot() routines now return an int. The return 
  value is 0 if the file could not be written, 1 otherwise.
- This version contains the alpha version of a perl5 module, which let's you
  access all the capabilities of the Vienna RNA library from perl scripts.

Version 1.2
-----------
- New energy parameters from (Walter et.al 1994). 
- Energy parameters can be read from file.
- RNAeval and energy_of_struct() support logarithmic energy function for 
  multi-loops.
- gmlRNA() produces secondary structure drawing in gml (Graph Meta Language). 
- Many bug fixes.


Progs/RNAcofold.c
/* Last changed Time-stamp: <2006-04-05 12:58:49 ivo> */
/*
		  c Ivo L Hofacker, Vienna RNA package
*/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <ctype.h>
#include <unistd.h>
#include <string.h>
#include "PS_dot.h"
#include "cofold.h"
#include "fold.h"
#include "part_func_co.h"
#include "part_func.h"
#include "fold_vars.h"
#include "utils.h"

extern void  read_parameter_file(const char fname[]);

/*@unused@*/
static char rcsid[] = "$Id: RNAcofold.c,v 1.7 2006/05/10 15:14:27 ivo Exp $";

#define PRIVATE static

static char  scale[] = "....,....1....,....2....,....3....,....4"
		       "....,....5....,....6....,....7....,....8";

PRIVATE char *costring(char *string);
PRIVATE char *tokenize(char *line);
PRIVATE void usage(void);
PRIVATE cofoldF do_partfunc(char *string, int length, int Switch, struct plist **tpr, struct plist **mf);
PRIVATE double *read_concentrations(FILE *fp);
PRIVATE void do_concentrations(double FEAB, double FEAA, double FEBB, double FEA, double FEB, double *startconces);
PRIVATE struct plist *get_mfe_plist(struct plist *pl);

/*--------------------------------------------------------------------------*/

int main(int argc, char *argv[])
{
  char *string, *line;
  char *structure=NULL, *cstruc=NULL;
  char  fname[53], ffname[60];
  char  *ParamFile=NULL;
  char  *ns_bases=NULL, *c;
  char *Concfile;
  int   i, length, l, sym, r;
  double min_en;
  double kT, sfact=1.07;
  int   pf=0, istty;
  int noconv=0;
  int doT=0;    /*compute dimere free energies etc.*/
  int doC=0;    /*toggle to compute concentrations*/
  int doQ=0;    /*toggle to compute prob of base being paired*/
  int cofi=0;   /*toggle concentrations stdin / file*/
  struct plist *prAB;
  struct plist *prAA;   /*pair probabilities of AA dimer*/
  struct plist *prBB;
  struct plist *prA;
  struct plist *prB;
  struct plist *mfAB;
  struct plist *mfAA;   /*pair mfobabilities of AA dimer*/
  struct plist *mfBB;
  struct plist *mfA;
  struct plist *mfB;
  double *ConcAandB;

  do_backtrack = 1;
  string=NULL;
  Concfile=NULL;
  for (i=1; i<argc; i++) {
    if (argv[i][0]=='-')
      switch ( argv[i][1] )
	{
	case 'T':  if (argv[i][2]!='\0') usage();
	  if(i==argc-1) usage();
	  r=sscanf(argv[++i], "%lf", &temperature);
	  if (!r) usage();
	  break;
	case 'p':  pf=1;
	  if (argv[i][2]!='\0')
	    (void) sscanf(argv[i]+2, "%d", &do_backtrack);
	  break;
	case 'n':
	  if ( strcmp(argv[i], "-noGU")==0) noGU=1;
	  if ( strcmp(argv[i], "-noCloseGU")==0) no_closingGU=1;
	  if ( strcmp(argv[i], "-noLP")==0) noLonelyPairs=1;
	  if ( strcmp(argv[i], "-nsp") ==0) {
	    if (i==argc-1) usage();
	    ns_bases = argv[++i];
	  }
	  if ( strcmp(argv[i], "-noconv")==0) noconv=1;
	  break;
	case '4':
	  tetra_loop=0;
	  break;
	case 'e':
	  if(i==argc-1) usage();
	  r=sscanf(argv[++i],"%d", &energy_set);
	  if (!r) usage();
	  break;
	case 'C':
	  fold_constrained=1;
	  break;
	case 'S':
	  if(i==argc-1) usage();
	  r=sscanf(argv[++i],"%lf", &sfact);
	  if (!r) usage();
	  break;
	case 'd': dangles=0;
	  if (argv[i][2]!='\0') {
	    r=sscanf(argv[i]+2, "%d", &dangles);
	    if (r!=1) usage();
	  }
	  break;
	case 'P':
	  if (i==argc-1) usage();
	  ParamFile = argv[++i];
	  break;
	case 'a':
	  doT=1;
	  pf=1;
	  break;
	case 'c':/*concentrations from stdin*/
	  doC=1;
	  doT=1;
	  pf=1;
	  break;
	case 'f':/*concentrations in file*/
	  if (i==argc-1) usage();
	  Concfile = argv[++i];
	  doC=1;
	  cofi=1;
	  doT=1;
	  pf=1;
	  break;
	case 'q':
	  pf=1;
	  doQ=1;
	  break;

	default: usage();
	}
  }

  if (ParamFile != NULL)
    read_parameter_file(ParamFile);

  if (ns_bases != NULL) {
    nonstandards = space(33);
    c=ns_bases;
    i=sym=0;
    if (*c=='-') {
      sym=1; c++;
    }
    while (*c!='\0') {
      if (*c!=',') {
	nonstandards[i++]=*c++;
	nonstandards[i++]=*c;
	if ((sym)&&(*c!=*(c-1))) {
	  nonstandards[i++]=*c;
	  nonstandards[i++]=*(c-1);
	}
      }
      c++;
    }
  }
  istty = isatty(fileno(stdout))&&isatty(fileno(stdin));
  if ((fold_constrained)&&(istty)) {
    printf("Input constraints using the following notation:\n");
    printf("| : paired with another base\n");
    printf(". : no constraint at all\n");
    printf("x : base must not pair\n");
    printf("< : base i is paired with a base j<i\n");
    printf("> : base i is paired with a base j>i\n");
    printf("matching brackets ( ): base i pairs base j\n");
  }

  do {				/* main loop: continue until end of file */
    cut_point = -1;
    if (istty) {
      printf("\nInput sequence(s); @ to quit\n");
      printf("Use '&' as spearator between 2 sequences that shall form a complex.\n");
      printf("%s\n", scale);
    }
    fname[0]='\0';
    if ((line = get_line(stdin))==NULL) break;

    /* skip comment lines and get filenames */
    while ((*line=='*')||(*line=='\0')||(*line=='>')) {
      if (*line=='>')
	(void) sscanf(line, ">%51s", fname);
      printf("%s\n", line);
      free(line);
      if ((line = get_line(stdin))==NULL) break;
    }

    if ((line ==NULL) || (strcmp(line, "@") == 0)) break;

    string = tokenize(line); /* frees line */

    length = (int) strlen(string);
    if (doC) {
      FILE *fp;
      if (cofi) { /* read from file */
	fp = fopen(Concfile, "r");
	if (fp==NULL) {
	  fprintf(stderr, "could not open concentration file %s", Concfile);
	  nrerror("\n");
	}
	ConcAandB = read_concentrations(fp);
	fclose(fp);
      } else {
	printf("Please enter concentrations [mol/l]\n format: ConcA ConcB\n return to end\n");
	ConcAandB = read_concentrations(stdin);
      }
    }

    structure = (char *) space((unsigned) length+1);
    if (fold_constrained) {
      cstruc = tokenize(get_line(stdin));
      if (cstruc!=NULL)
	strncpy(structure, cstruc, length);
      else
	fprintf(stderr, "constraints missing\n");
    }

    for (l = 0; l < length; l++) {
      string[l] = toupper(string[l]);
      if (!noconv && string[l] == 'T') string[l] = 'U';
    }
    if (istty) {
      if (cut_point == -1)
	printf("length = %d\n", length);
      else
	printf("length1 = %d\nlength2 = %d\n",
	       cut_point-1, length-cut_point+1);
    }

    /*compute mfe of AB dimer*/
    min_en = cofold(string, structure);
    mfAB=(struct plist *) space(sizeof(struct plist) * (length+1));
    mfAB=get_mfe_plist(mfAB);

    if (cut_point == -1)    printf("%s\n%s", string, structure); /*no cofold*/

    else {
      char *pstring, *pstruct;
      pstring = costring(string);
      pstruct = costring(structure);
      printf("%s\n%s", pstring,  pstruct);
      free(pstring);
      free(pstruct);
    }
    if (istty)
      printf("\n minimum free energy = %6.2f kcal/mol\n", min_en);
    else
      printf(" (%6.2f)\n", min_en);

      (void) fflush(stdout);

    if (fname[0]!='\0') {
      strcpy(ffname, fname);
      strcat(ffname, "_ss.ps");
    } else {
      strcpy(ffname, "rna.ps");
    }
    if (length<2000)
      (void) PS_rna_plot(string, structure, ffname);
    else {
     fprintf(stderr,"INFO: structure too long, not doing xy_plot\n");
     free_co_arrays();
    }

    /*compute partition function*/
    if (pf) {
      cofoldF AB, AA, BB;
      if (dangles==1) {
	dangles=2;   /* recompute with dangles as in pf_fold() */
	min_en = energy_of_struct(string, structure);
	dangles=1;
      }

      kT = (temperature+273.15)*1.98717/1000.; /* in Kcal */
      pf_scale = exp(-(sfact*min_en)/kT/length);
      if (length>2000) fprintf(stderr, "scaling factor %f\n", pf_scale);

      init_co_pf_fold(length);

      if (cstruc!=NULL)
	strncpy(structure, cstruc, length+1);
      AB = co_pf_fold(string, structure);

      if (do_backtrack) {
	char *costruc;
	costruc = (char *) space(sizeof(char)*(strlen(structure)+2));
	if (cut_point<0) printf("%s", structure);
	else {
	  strncpy(costruc, structure, cut_point-1);
	  strcat(costruc, "&");
	  strcat(costruc, structure+cut_point-1);
	  printf("%s", costruc);
	}
	if (!istty) printf(" [%6.2f]\n", AB.FAB);
	else printf("\n");/*8.6.04*/
      }
      if ((istty)||(!do_backtrack))
	printf(" free energy of ensemble = %6.2f kcal/mol\n", AB.FAB);
      printf(" frequency of mfe structure in ensemble %g",
	     exp((AB.FAB-min_en)/kT));

      printf(" , delta G binding=%6.2f\n", AB.FcAB - AB.FA - AB.FB);

      prAB=(struct plist *) space(sizeof(struct plist) * (2*length));
      prAB=get_plist(prAB, length,0.00001);

      /* if (doQ) make_probsum(length,fname); */ /*compute prob of base paired*/
      /* free_co_arrays(); */
      if (doT) { /* cofold of all dimers, monomers */
	int Blength, Alength;
	char  *Astring, *Bstring;
	char *Newstring;
	char Newname[30];
	char comment[80];
	if (cut_point<0) {
	  printf("Sorry, i cannot do that with only one molecule, please give me two or leave it\n");
	  free(mfAB);
	  free(prAB);
	  continue;
	}
	if (dangles==1) dangles=2;
	Alength=cut_point-1;        /*length of first molecule*/
	Blength=length-cut_point+1; /*length of 2nd molecule*/

	Astring=(char *)space(sizeof(char)*(Alength+1));/*Sequence of first molecule*/
	Bstring=(char *)space(sizeof(char)*(Blength+1));/*Sequence of second molecule*/
	strncat(Astring,string,Alength);
	strncat(Bstring,string+Alength,Blength);

	/* compute AA dimer */
	prAA=(struct plist *) space(sizeof(struct plist) * (4*Alength));
	mfAA=(struct plist *) space(sizeof(struct plist) * (Alength+1));
	AA=do_partfunc(Astring, Alength, 2, &prAA, &mfAA);
	/* compute BB dimer */
	prBB=(struct plist *) space(sizeof(struct plist) * (4*Blength));
	mfBB=(struct plist *) space(sizeof(struct plist) * (Blength+1));
	BB=do_partfunc(Bstring, Blength, 2, &prBB, &mfBB);
	/*free_co_pf_arrays();*/

	/* compute A monomer */
	prA=(struct plist *) space(sizeof(struct plist) * (2*Alength));
	mfA=(struct plist *) space(sizeof(struct plist) * (Alength+1));
	do_partfunc(Astring, Alength, 1, &prA, &mfA);

	/* compute B monomer */
	prB=(struct plist *) space(sizeof(struct plist) * (2*Blength));
	mfB=(struct plist *) space(sizeof(struct plist) * (Blength+1));
	do_partfunc(Bstring, Blength, 1, &prB, &mfB);

	compute_probabilities(AB.F0AB, AB.FA, AB.FB, prAB, prA, prB, Alength);
	compute_probabilities(AA.F0AB, AA.FA, AA.FA, prAA, prA, prA, Alength);
	compute_probabilities(BB.F0AB, BB.FA, BB.FA, prBB, prA, prB, Blength);
	printf("Free Energies:\nAB\t\tAA\t\tBB\t\tA\t\tB\n%.6f\t%6f\t%6f\t%6f\t%6f\n",
	       AB.FcAB, AA.FcAB, BB.FcAB, AB.FA, AB.FB);

	if (doC) {
	  do_concentrations(AB.FcAB, AA.FcAB, BB.FcAB, AB.FA, AB.FB, ConcAandB);
	  free(ConcAandB);/*freeen*/
	}

	if (fname[0]!='\0') {
	  strcpy(ffname, fname);
	  strcat(ffname, "_dp5.ps");
	} else strcpy(ffname, "dot5.ps");
	/*output of the 5 dot plots*/

	/*AB dot_plot*/
	/*write Free Energy into comment*/
	sprintf(comment,"\n%%Heterodimer AB FreeEnergy= %.9f\n", AB.FcAB);
	/*reset cut_point*/
	cut_point=Alength+1;
	/*write New name*/
	strcpy(Newname,"AB");
	strcat(Newname,ffname);
	(void)PS_dot_plot_list(string, Newname, prAB, mfAB, comment);

	/*AA dot_plot*/
	sprintf(comment,"\n%%Homodimer AA FreeEnergy= %.9f\n",AA.FcAB);
	/*write New name*/
	strcpy(Newname,"AA");
	strcat(Newname,ffname);
	/*write AA sequence*/
	Newstring=(char*)space((2*Alength+1)*sizeof(char));
	strcpy(Newstring,Astring);
	strcat(Newstring,Astring);
	(void)PS_dot_plot_list(Newstring, Newname, prAA, mfAA, comment);
	free(Newstring);

	/*BB dot_plot*/
	sprintf(comment,"\n%%Homodimer BB FreeEnergy= %.9f\n",BB.FcAB);
	/*write New name*/
	strcpy(Newname,"BB");
	strcat(Newname,ffname);
	/*write BB sequence*/
	Newstring=(char*)space((2*Blength+1)*sizeof(char));
	strcpy(Newstring,Bstring);
	strcat(Newstring,Bstring);
	/*reset cut_point*/
	cut_point=Blength+1;
	(void)PS_dot_plot_list(Newstring, Newname, prBB, mfBB, comment);
	free(Newstring);

	/*A dot plot*/
	/*reset cut_point*/
	cut_point=-1;
	sprintf(comment,"\n%%Monomer A FreeEnergy= %.9f\n",AB.FA);
	/*write New name*/
	strcpy(Newname,"A");
	strcat(Newname,ffname);
	/*write BB sequence*/
	(void)PS_dot_plot_list(Astring, Newname, prA, mfA, comment);

	/*B monomer dot plot*/
	sprintf(comment,"\n%%Monomer B FreeEnergy= %.9f\n",AB.FB);
	/*write New name*/
	strcpy(Newname,"B");
	strcat(Newname,ffname);
	/*write BB sequence*/
	(void)PS_dot_plot_list(Bstring, Newname, prB, mfB, comment);
	free(Astring); free(Bstring); free(prAB); free(prAA); free(prBB); free(prA); free(prB);
	free(mfAB); free(mfAA); free(mfBB); free(mfA); free(mfB);

      } /*end if(doT)*/

    }/*end if(pf)*/


    if (do_backtrack) {
      if (fname[0]!='\0') {
	strcpy(ffname, fname);
	strcat(ffname, "_dp.ps");
      } else strcpy(ffname, "dot.ps");

      if (!doT) {
	if (pf) {	  (void) PS_dot_plot_list(string, ffname, prAB, mfAB, "doof");
	free(prAB);}
	free(mfAB);
      }
    }
    if (!doT) free_co_pf_arrays();


    if (cstruc!=NULL) free(cstruc);
    (void) fflush(stdout);
    free(string);
    free(structure);
  } while (1);
  return 0;
}

PRIVATE char *tokenize(char *line)
{
  char *pos, *copy;
  int cut = -1;

  copy = (char *) space(strlen(line)+1);
  (void) sscanf(line, "%s", copy);
  pos = strchr(copy, '&');
  if (pos) {
    cut = (int) (pos-copy)+1;
    if (cut >= strlen(copy)) cut = -1;
    if (strchr(pos+1, '&')) nrerror("more than one cut-point in input");
    for (;*pos;pos++) *pos = *(pos+1); /* splice out the & */
  }
  if (cut > -1) {
    if (cut_point==-1) cut_point = cut;
    else if (cut_point != cut) {
      fprintf(stderr,"cut_point = %d cut = %d\n", cut_point, cut);
      nrerror("Sequence and Structure have different cut points.");
    }
  }
  free(line);
  return copy;
}

PRIVATE char *costring(char *string)
{
  char *ctmp;
  int len;

  len = strlen(string);
  ctmp = (char *)space((len+2) * sizeof(char));
  /* first sequence */
  (void) strncpy(ctmp, string, cut_point-1);
  /* spacer */
  ctmp[cut_point-1] = '&';
  /* second sequence */
  (void) strcat(ctmp, string+cut_point-1);

  return ctmp;
}

PRIVATE cofoldF do_partfunc(char *string, int length, int Switch, struct plist **tpr, struct plist **mfpl) {
  /*compute mfe and partition function of dimere or  monomer*/
  char *Newstring;
  char *tempstruc;
  double min_en;
  double sfact=1.07;
  double kT;
  cofoldF X;
  kT = (temperature+273.15)*1.98717/1000.;
  switch (Switch)
    {
    case 1: /*monomer*/
      cut_point=-1;
      tempstruc = (char *) space((unsigned)length+1);
      min_en = fold(string, tempstruc);
      pf_scale = exp(-(sfact*min_en)/kT/(length));
      *mfpl=get_mfe_plist(*mfpl);
      free_arrays();
      /*En=pf_fold(string, tempstruc);*/
      init_co_pf_fold(length);
      X=co_pf_fold(string, tempstruc);

      *tpr=get_plist(*tpr, length,0.00001);
      free_co_pf_arrays();
      free(tempstruc);
      break;

    case 2: /*dimer*/
      Newstring=(char *)space(sizeof(char)*(length*2+1));
      strcat(Newstring,string);
      strcat(Newstring,string);
      cut_point=length+1;
      tempstruc = (char *) space((unsigned)length*2+1);
      min_en = cofold(Newstring, tempstruc);
      pf_scale =exp(-(sfact*min_en)/kT/(2*length));
      *mfpl=get_mfe_plist(*mfpl);
      free_co_arrays();
      init_co_pf_fold(2*length);
      X=co_pf_fold(Newstring, tempstruc);
      *tpr=get_plist(*tpr, 2*length,0.00001);
      free_co_pf_arrays();
      free(Newstring);
      free(tempstruc);
      break;

    default:
      printf("Error in get_partfunc\n, computing neither mono- nor dimere!\n");
      exit (42);

    }
  return X;
}


PRIVATE void do_concentrations(double FEAB, double FEAA, double FEBB, double FEA, double FEB, double *startconc) {
  /* compute and print concentrations out of free energies, calls get_concentrations */
  struct ConcEnt *result;
  int i, n;

  result=get_concentrations(FEAB, FEAA, FEBB, FEA, FEB, startconc);

  printf("Initial concentrations\t\trelative Equilibrium concentrations\n");
  printf("A\t\t B\t\t AB\t\t AA\t\t BB\t\t A\t\t B\n");
  for (n=0; (startconc[2*n]>0) || (startconc[2*n+1]>0); n++); /* count */
  for (i=0; i<n ;i++) {
    double tot = result[i].A0+result[i].B0;
    printf("%-10g\t%-10g\t%.5f \t%.5f \t%.5f \t%.5f \t%.5f\n", result[i].A0, result[i].B0,
	   result[i].ABc/tot, result[i].AAc/tot, result[i].BBc/tot ,result[i].Ac/tot, result[i].Bc/tot);
  }
  free(result);
}


PRIVATE double *read_concentrations(FILE *fp) {
  /* reads concentrations, returns list of double, -1. marks end */
  char *line;
  double *startc;
  int i=0, n=2;

  startc = (double *) space((2*n+1)*sizeof(double));

  while ((line=get_line(fp))!=NULL) {
    int c;
    if (i+4>=2*n) {
      n*=2;
      startc=(double *)xrealloc(startc,(2*n+1)*sizeof(double));
    }
    c = sscanf(line,"%lf %lf", &startc[i], &startc[i+1]);
    free(line);
    if (c<2) break;
    i+=2;
  }
  startc[i]=startc[i+1]=0;
  return startc;
}

static struct plist *get_mfe_plist(struct plist *pl) {
  /*get list of mfe structure out of base_pairs array*/
  int l;
  for(l=1; l<=base_pair[0].i; l++)
    {
      pl[l-1].i=base_pair[l].i;
      pl[l-1].j=base_pair[l].j;
      pl[l-1].p=0.95;
    }
  pl[l-1].i=0;
  pl[l-1].j=0;
  pl[l-1].p=0.;
  pl=(struct plist *)xrealloc(pl,l*sizeof(struct plist));
  return pl;
}


PRIVATE void usage(void)
{
  nrerror("usage:\n"
	  "RNAcofold [-a] [-c] [-f concfile]\n"
	  "[-p[0]] [-C] [-T temp] [-4] [-d[2|3]] [-noGU] [-noCloseGU]\n"
	  "[-noLP] [-e e_set] [-P paramfile] [-nsp pairs] [-S scale] "
	  "[-noconv]\n");
}
